### PR TITLE
feat: Integrate vLLM with Weight Propagation Interface (WPI) for Zero-Copy Weight Transfer

### PR DESCRIPTION
**Pull request by yangspirit**
## Purpose
To integrate vLLM with the **[Weight Propagation Interface (WPI)](https://github.com/llm-d-incubation/weight-propagation-interface)**, a Kubernetes-native framework for high-speed, zero-copy movement of large model weights. This feature enables zero-copy weight sharing from external trainers directly to vLLM inference workers, eliminating storage and CPU-GPU copy bottlenecks during online reinforcement learning (e.g., with GRPO).

This PR adds:

A new WPIWeightTransferEngine registered in the factory.
Lazy loading support for the required wpi_client package.

## Test Plan
The integration was tested in a GKE cluster on accelerator-equipped nodes with the WPI driver running.

To reproduce the test:

Deploy the WPI driver and Custom Resources (WeightBuffer/WeightClaim).
Deploy the vLLM server with the "wpi" backend configured.
Run the cluster benchmark script:
bash
cd weight-propagation-interface/consumer/wpi_vllm/benchmark
./run_cluster_benchmark.sh Qwen/Qwen2-7B
Verify that the trainer can successfully connect and perform zero-copy weight transfers.

## Test Result
Verified using a Qwen/Qwen2-7B model (approx. 14.19 GB weight payload) in a single-GPU setup:

Aggregate Mean Bandwidth: ~20.43 GB/s (peaking at 20.61 GB/s).
Mean Total Time: ~694.46 ms for complete weight propagation and HTTP metadata delivery.
The server pod remained stable without restarts, successfully loading the model and capturing CUDA graphs.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+    try:
+        from wpi_client.client import WPIClient
+        return WPIClient
+    except ImportError:
+        raise ImportError(
+            "WPI weight transfer backend requires the `wpi_client` package. "
+            "Install it with: pip install wpi_client\n"
+            "Or from the WPI source: cd weight-propagation-interface/consumer/"
+            "wpi_client && pip install -e ."
+        ) from None
```
Comment (SumanthRH) [inline vllm/distributed/weight_transfer/wpi_engine.py:231]: It looks like the package is not yet on PyPI. Do we want to hold off on the PR until the package is published?

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+    buffer_id: str
+    buffer_size_bytes: int
+    target_node_ids: list[str]
+
+
+@dataclass
+class WPITrainerSendWeightsArgs:
+    """Arguments for WPI trainer_send_weights method."""
+
+    mode: str
```
Comment (SumanthRH) [inline vllm/distributed/weight_transfer/wpi_engine.py]: Just FYI we are trying to clean up the usage of `mode` in the weight transfer APIs.  #37476 is renaming this to `send_mode` to be explicit. (This is a pretty clunky part of the code that we want to remove long term, but thought I'll mention the planned change short term)

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+
+    shard_index: int = -1
+    """Shard index for tensor-parallel (-1 = unsharded)."""
+
+    total_shards: int = 0
+    """Total number of shards (0 = unsharded)."""
+
+    # Pre-initialized trainer context (from trainer_init())
+    trainer_ctx: WPITrainerContext | None = None
+    """Pre-initialized trainer context. If None, one will be created."""
```
Comment (SumanthRH) [inline vllm/distributed/weight_transfer/wpi_engine.py:126]: Just for my understanding, when would callers use this pre-initialized context vs not?

**Review by SumanthRH — COMMENTED** [review]
Very clean! Could you add an example in `examples/rl` for weight transfer with WPI? Perhaps similar to this example: https://github.com/vllm-project/vllm/blob/597ed138033e51355ff4ba49876578df92633c91/examples/rl/rlhf_http_nccl.py

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 1 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:231]: Yes, the package is not on PyPI yet. I have updated the error message in wpi_engine.py to remove the PyPI reference and only provide instructions on how to install it directly from the WPI source (cd weight-propagation-interface/consumer/wpi_client && pip install -e .).

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 2 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py]: Thanks for the heads-up! Makes sense to make it more explicit. I've kept it as mode for now in the example to match the current WPITrainerSendWeightsArgs definition in this branch.

I'll keep an eye on #37476 and will update it to send_mode once that lands.

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 3 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:126]: Callers should use the pre-initialized context when performing repeated weight transfers (like in an RLHF training loop). Initializing it once at the start avoids the overhead of staging the buffer and importing CUDA memory on every step.

They would omit it (relying on lazy init) for one-off transfers or simple testing where they want to avoid managing the context object and don't care about the initialization overhead.

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 2 above)
Comment (SumanthRH) [reply vllm/distributed/weight_transfer/wpi_engine.py]: @yangspirit `send_mode` has landed. Can we update the PR to be consistent?

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 3 above)
Comment (SumanthRH) [reply vllm/distributed/weight_transfer/wpi_engine.py:126]: Good to know, thanks!

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 2 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py]: Done. PTAL.

**Review by SumanthRH — COMMENTED** [review]
can we also add a documentation page for the WPI engine?

At `docs/training/weight_transfer`

**Comment by aoshen02** [comment]
Hi, thanks for the contribution, do you mind adding more experiment result in at least a 8-gpu rack and compare wpi with nccl and ipc?

**Comment by yangspirit** [comment]
> Hi, thanks for the contribution, do you mind adding more experiment result in at least a 8-gpu rack and compare wpi with nccl and ipc?

Done.

**Comment by aoshen02** [comment]
> > Hi, thanks for the contribution, do you mind adding more experiment result in at least a 8-gpu rack and compare wpi with nccl and ipc?
> 
> Done.

I've enabled the CI

**Comment by aoshen02** [comment]
I think because this PR lasts too long, so there are some problems.
This class does not implement `start_weight_update()` or
`finish_weight_update()`. Since both are abstract in `WeightTransferEngine`,
the WPI backend cannot be instantiated by the factory.

Please add both methods, likely matching the lifecycle used by the NCCL and IPC
engines.

**Comment by yangspirit** [comment]
> I think because this PR lasts too long, so there are some problems. This class does not implement `start_weight_update()` or `finish_weight_update()`. Since both are abstract in `WeightTransferEngine`, the WPI backend cannot be instantiated by the factory.
> 
> Please add both methods, likely matching the lifecycle used by the NCCL and IPC engines.

Done. PTAL.

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+"""WPI-based weight transfer engine using the Weight Propagation Interface.
+
+The WPI (Weight Propagation Interface) driver manages a pre-allocated, persistent
+VRAM buffer on each node and uses NCCL broadcast (NodePropagate) to distribute
+weights across nodes at InfiniBand/RDMA speeds. This engine integrates WPI into
+vLLM's WeightTransferEngine framework, providing zero-copy weight updates for
+RLHF and online training workflows.
+
+Architecture:
+    Trainer (send side)                 vLLM Worker (receive side)
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py:12]: should follow start, for loop, finish contract here.

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+            dtype_names.append(str(weight.dtype).split(".")[-1])
+            shapes.append(list(weight.shape))
+            offsets.append(offset)
+
+            # Copy into the shared VRAM buffer
+            ctx.vram_buffer[offset : offset + nbytes].copy_(
+                weight.view(-1).view(torch.uint8),
+                non_blocking=True,
+            )
+            offset += nbytes
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py:686]: Codex found a problem I feel like make a lot of sense:
The buffer is byte-addressable, but the receiver reinterprets each slice with .view(dtype=dtype) without copying it into a new storage. Therefore, the recorded offset must be divisible by the target dtype’s itemsize.
For example, a 6-byte FP16 tensor followed by an FP32 tensor places the FP32 tensor at offset 6. The receiver then attempts buffer[6:...].view(torch.float32), which fails because the underlying storage offset is not divisible by 4.
Please align offset before recording and copying each tensor, and include the resulting padding when calculating buffer_size_bytes, for example:
        for name, tensor in iterator:
            weight = tensor.detach().contiguous()
            itemsize = weight.element_size()
            offset = (offset + itemsize - 1) // itemsize * itemsize
            nbytes = weight.nbytes
The example’s total_model_bytes = sum(...) must also use the same aligned-layout calculation; otherwise the aligned final tensor can exceed the allocated buffer.

**Comment by aoshen02** [comment]
@codex  review

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 4 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:12]: Done.

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 5 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:686]: Done

File: examples/rl/rlhf_http_wpi.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    print("Propagating weights via WPI...")
+    param_iter = ((n, p) for n, p in train_model.named_parameters())
+    args = WPITrainerSendWeightsArgs(
+        send_mode="http",
+        url=BASE_URL,
+        trainer_ctx=ctx,
+    )
+
+    # This will pack weights, trigger WPI propagate, and call /update_weights on server
+    WPIWeightTransferEngine.trainer_send_weights(param_iter, args)
```
[the remark this answers is not in the document]
Comment (yangspirit) [reply examples/rl/rlhf_http_wpi.py:197]: Done

File: docs/training/weight_transfer/wpi.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+To perform a weight update, the trainer packs model parameters into the mapped WPI buffer, triggers the WPI driver's NCCL propagation, and sends the layout metadata to vLLM workers (via HTTP or Ray control plane):
+
+```python
+from vllm.distributed.weight_transfer.wpi_engine import (
+    WPIWeightTransferEngine,
+    WPITrainerSendWeightsArgs,
+)
+
+# Start weight update on inference side
+llm.start_weight_update(is_checkpoint_format=True)
```
[the remark this answers is not in the document]
Comment (yangspirit) [reply docs/training/weight_transfer/wpi.md]: Done

**Comment by aoshen02** [comment]
@codex review

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+            buffer_id,
+            buffer_size_bytes,
+            target_node_ids,
+        )
+
+        return WPITrainerContext(
+            client=client,
+            vram_buffer=vram_buffer,
+            buffer_id=buffer_id,
+            buffer_size_bytes=buffer_size_bytes,
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py:583]: I found a codex comment make sense to me, could you check?

stage_weight() above receives shard_index and total_shards, and WPI registers a sharded buffer under the effective ID <buffer_id>__shard_<N>. However, the returned context only preserves the base buffer_id, discarding the shard identity that is needed by later operations.
Consequently, trainer_send_weights() calls propagate(buffer_id=ctx.buffer_id) with weights, while the staged source resource is actually weights__shard_N, so the driver cannot find the buffer.
Please store shard_index/total_shards or the effective buffer ID in WPITrainerContext, and use that same resource identity for propagation and cleanup.

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+        3. Receives the CUDA memory FD via SCM_RIGHTS UNIX socket
+        4. Imports the FD as mapped CUDA memory (cuMemImportFromShareableHandle)
+        5. Connects to the notify socket for READY signals
+
+        Args:
+            init_info: WPI initialization info with buffer ID, size, and driver addr
+        """
+        WPIClient = _import_wpi_client()
+
+        self._buffer_id = init_info.buffer_id
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py:297]: Is it ok that for every tp rank use the same buffer id?

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+            socket_dir=init_info.socket_dir,
+            driver_port=init_info.driver_port,
+        )
+
+        # Stage the empty receive buffer on the local WPI driver
+        if not self._staged:
+            self._client.stage_weight(
+                buffer_id=self._buffer_id,
+                size_bytes=self._buffer_size,
+                claim_id=f"{self._buffer_id}-claim",
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py]: And claim_id is the same, seems it's related to shutdown and resource release

**Comment by aoshen02** [comment]
Also find a codex comment make sense:
WPI currently propagates the buffer as size_bytes // 2 FP16 elements. If buffer_size_bytes is odd, the last byte is not transferred, which can silently corrupt UINT8/FP8 weights.
Please round the physical buffer size up to an even value on both trainer and worker sides, or reject unaligned sizes until WPI supports byte-level transfers.

**Comment by aoshen02** [comment]
Could you run an E2E test after these things been addressed?

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+                shard_index=shard_index,
+                total_shards=total_shards,
```
[the remark this answers is not in the document]
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:352]: Done.

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 8 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:583]: Done.

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 9 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py:297]: Done

File: vllm/distributed/weight_transfer/wpi_engine.py (hunk 10 above)
Comment (yangspirit) [reply vllm/distributed/weight_transfer/wpi_engine.py]: Done

**Comment by yangspirit** [comment]
> Could you run an E2E test after these things been addressed?

I have addressed the comments and ran the E2E tests. Thanks for your review!

**Comment by aoshen02** [comment]
> > Could you run an E2E test after these things been addressed?
> 
> I have addressed the comments and ran the E2E tests. Thanks for your review!

thanks u!

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+                        device_index
+                    )
+                )
+            except Exception:
+                absolute_gpu_id = device_index
+
+            fd = self._client.receive_fd(
+                self._buffer_id,
+                gpu_id=absolute_gpu_id,
+                shard_index=shard_index,
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py:351]: Could you check this codex code review?  It seems problem still exits with your update
Passing absolute_gpu_id only ensures that WPI creates the relocated buffer on the correct GPU. The current WPI driver copies that buffer once during receive_fd(), while later NodePropagate calls update only the original allocation.
Therefore, workers on non-owner GPUs can still receive READY and load stale weights. Please either update every relocated buffer after each propagation, stage and propagate a separate buffer per GPU, or reject multi-GPU-per-node configurations.
WPI currently performs the one-time relocation [here](https://github.com/llm-d-incubation/weight-propagation-interface/blob/e8db68a666e292646ab32b7273d75e59c61bc4dc/driver/main.py#L364-L413), while propagation writes only the original allocation [here](https://github.com/llm-d-incubation/weight-propagation-interface/blob/e8db68a666e292646ab32b7273d75e59c61bc4dc/driver/main.py#L661-L783).

**Comment by aoshen02** [comment]
What end to end test do you test on, is it an internal rl framework? Could you test with tp and ep enabled? non-colocate case would be better if covered. Could you make a reproducable example that I can test it on?

File: vllm/distributed/weight_transfer/wpi_engine.py [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+    buffer_id: str
+    buffer_size_bytes: int
+    target_node_ids: list[str]
+    effective_buffer_id: str = ""
+    claim_id: str = ""
+
+    def close(self) -> None:
+        """Release the staged trainer VRAM buffer and close the client."""
+        if self.client is not None:
+            if self.claim_id:
```
Comment (aoshen02) [inline vllm/distributed/weight_transfer/wpi_engine.py:87]: dead code?
